src/ring_theory/polynomial/eisenstein.lean

```diff
@@ -0,0 +1,187 @@
+/-
+Copyright (c) 2022 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import ring_theory.eisenstein_criterion
+
+/-!
+# Eisenstein polynomials
+Given an ideal `𝓟` of a commutative semiring `R`, we say that a polynomial `f : polynomial R` is
+*Eisenstein at `𝓟`* if `f.leading_coeff ∉ 𝓟`, `∀ n, n < f.nat_degree → f.coeff n ∈ 𝓟` and
+`f.coeff 0 ∉ 𝓟 ^ 2`. In this file we gather miscellaneous results about Eisenstein polynomials.
+
+## Main definitions
+* `polynomial.is_eisenstein_at f 𝓟` : the property of being Eisenstein at `𝓟`.
+
+## Main results
+* `polynomial.is_eisenstein_at.irreducible` : if a primitive `f` satisfies `f.is_eisenstein_at 𝓟`,
+  where `𝓟.is_prime`, then `f` is irreducible.
+
+## Implementation details
+We also define a notion `is_weakly_eisenstein_at` requiring only that
+`∀ n < f.nat_degree → f.coeff n ∈ 𝓟`. This makes certain results slightly more general and it is
+useful since it is sometimes better behaved (for example it is stable under `polynomial.map`).
+
+-/
+
+universes u v
+
+variables {R : Type u}
+
+open ideal algebra finset
+
+open_locale big_operators
+
+namespace polynomial
+
+/-- Given an ideal `𝓟` of a commutative semiring `R`, we say that a polynomial `f : polynomial R`
+is *weakly Eisenstein at `𝓟`* if `∀ n, n < f.nat_degree → f.coeff n ∈ 𝓟`. -/
+@[mk_iff] structure is_weakly_eisenstein_at [comm_semiring R] (f : polynomial R) (𝓟 : ideal R) :
+  Prop := (mem : ∀ {n}, n < f.nat_degree → f.coeff n ∈ 𝓟)
+
+/-- Given an ideal `𝓟` of a commutative semiring `R`, we say that a polynomial `f : polynomial R`
+is *Eisenstein at `𝓟`* if `f.leading_coeff ∉ 𝓟`, `∀ n, n < f.nat_degree → f.coeff n ∈ 𝓟` and
+`f.coeff 0 ∉ 𝓟 ^ 2`. -/
+@[mk_iff] structure is_eisenstein_at [comm_semiring R] (f : polynomial R) (𝓟 : ideal R) : Prop :=
+(leading : f.leading_coeff ∉ 𝓟)
+(mem : ∀ {n}, n < f.nat_degree → f.coeff n ∈ 𝓟)
+(not_mem : f.coeff 0 ∉ 𝓟 ^ 2)
+
+namespace is_weakly_eisenstein_at
+
+section comm_semiring
+
+variables [comm_semiring R] {𝓟 : ideal R} {f : polynomial R} (hf : f.is_weakly_eisenstein_at 𝓟)
+
+include hf
+
+lemma map {A : Type v} [comm_ring A] (φ : R →+* A) : (f.map φ).is_weakly_eisenstein_at (𝓟.map φ) :=
+begin
+  refine (is_weakly_eisenstein_at_iff _ _).2 (λ n hn, _),
+  rw [coeff_map],
+  exact mem_map_of_mem _ (hf.mem (lt_of_lt_of_le hn (nat_degree_map_le _ _)))
+end
+
+end comm_semiring
+
+section comm_ring
+
+variables [comm_ring R] {𝓟 : ideal R} {f : polynomial R} (hf : f.is_weakly_eisenstein_at 𝓟)
+variables {S : Type v} [comm_ring S] [algebra R S]
+
+section principal
+
+variable {p : R}
+
+local notation `P` := submodule.span R {p}
+
+lemma ex_mem_adjoin_mul_eq_pow_nat_degree {x : S} (hx : aeval x f = 0)
```

Comment:
I'm happy with abbreviation `exists` to `ex`, but it deviates from the style guide. Has this use of `ex` come up before? Should we adapt the style guide?

Reply:
Looking at `\b ex_` (I hope I didn't messed up with regular expressions...) confirm you're right, we don't use `ex` for `exist`.